nestkernel/eprop_archiving_node.cpp

```diff
@@ -0,0 +1,253 @@
+ *
+ *  NEST is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  NEST is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with NEST.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+// nestkernel
+#include "eprop_archiving_node.h"
+#include "kernel_manager.h"
+
+// sli
+#include "dictutils.h"
+
+namespace nest
+{
+
+nest::EpropArchivingNode::EpropArchivingNode()
+  : Node()
+{
+}
+
+nest::EpropArchivingNode::EpropArchivingNode( const EpropArchivingNode& n )
+  : Node( n )
+{
+}
+
+void
+nest::EpropArchivingNode::init_update_history()
+{
+  // register first entry for every synapse, increase access counter if entry already in list
+
+  long shift = get_shift();
+
+  std::vector< HistEntryEpropUpdate >::iterator it =
+    std::lower_bound( update_history_.begin(), update_history_.end(), shift );
+
+  if ( it == update_history_.end() or it->t_ != shift )
+    update_history_.insert( it, HistEntryEpropUpdate( shift, 1 ) );
+  else
+    ++it->access_counter_;
+}
+
+void
+nest::EpropArchivingNode::write_update_to_history( long t_previous_update, long t_current_update )
+{
+  long shift = get_shift();
+
+  std::vector< HistEntryEpropUpdate >::iterator it;
+
+  it = std::lower_bound( update_history_.begin(), update_history_.end(), t_current_update + shift );
+
+  if ( it != update_history_.end() or it->t_ == ( t_current_update + shift ) )
+    ++it->access_counter_;
+  else
+    update_history_.insert( it, HistEntryEpropUpdate( t_current_update + shift, 1 ) );
+
+  it = std::lower_bound( update_history_.begin(), update_history_.end(), t_previous_update + shift );
+
+  if ( it != update_history_.end() or it->t_ == ( t_previous_update + shift ) )
+    --it->access_counter_;
+}
+
+void
+nest::EpropArchivingNode::write_eprop_parameter_to_map( std::string parameter_name, double parameter_value )
+{
+  eprop_parameter_map_[ parameter_name ] = parameter_value;
+}
+
+void
+nest::EpropArchivingNode::write_surrogate_gradient_to_history( long time_step, double surrogate_gradient )
+{
+  eprop_history_.push_back( HistEntryEpropArchive( time_step, surrogate_gradient, 0.0 ) );
+}
+
+void
+nest::EpropArchivingNode::write_error_signal_to_history( long time_step, double error_signal )
+{
+  long shift = delay_out_norm_;
+  eprop_history_.push_back( HistEntryEpropArchive( time_step - shift, 0.0, error_signal ) );
+}
+
+void
+nest::EpropArchivingNode::write_learning_signal_to_history( long& time_step,
+  long& delay_out_rec,
+  double& weight,
+  double& error_signal )
```

Comment:
Reference do not make sense here; to tell the compiler that values need not be copied into the function, I'd suggest making the parameters const:
```suggestion
nest::EpropArchivingNode::write_learning_signal_to_history( const long time_step,
  const long delay_out_rec,
  const double weight,
  const double error_signal )
```

Reply:
Fixed in https://github.com/nest/nest-simulator/pull/2867/commits/87a8062d34b92f78bea73d9fbcba400d3436e7da.